test/infrastructure/docker/e2e/docker_test.go

```diff
@@ -0,0 +1,196 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"fmt"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/cluster-api/bootstrap/kubeadm/types/v1beta1"
+	"sigs.k8s.io/cluster-api/test/framework"
+
+	capiv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	bootstrapv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1alpha2"
+	infrav1 "sigs.k8s.io/cluster-api/test/infrastructure/docker/api/v1alpha2"
+)
+
+var _ = Describe("Docker", func() {
+	Describe("Cluster Creation", func() {
+
+		namespace := "default"
+
+		clusterGen := &ClusterGenerator{}
+		nodeGen := &NodeGenerator{}
+
+		Context("One node cluster", func() {
+			It("should create a single node cluster", func() {
+				cluster, infraCluster := clusterGen.GenerateCluster(namespace)
+				singleNode := nodeGen.GenerateNode(cluster.GetName())
+				input := &framework.OneNodeClusterInput{
+					Management:    mgmt,
+					Cluster:       cluster,
+					InfraCluster:  infraCluster,
+					Node:          singleNode,
+					CreateTimeout: 3 * time.Minute,
+				}
+
+				framework.OneNodeCluster(input)
+
+				cleanupInput := &framework.CleanUpInput{
+					Management: mgmt,
+					Cluster:    cluster,
+				}
+
+				framework.CleanUp(cleanupInput)
```

Comment:
Are we not doing a `defer` or `AfterEach` on the clean up step because we want the environment to remain for troubleshooting purposes in case something goes wrong in `framework.OneNodeCluster(input)`?
Also, if a single test block fails it could pollute the namespace for the following tests if not cleaned up.

Reply:
The namespace won't be polluted because the tests share the same generator and each cluster would have a unique name in the namespace.

For now, leaving the clusters around on failure has been easier to debug, but if it turns out we really need these to be in defer/afterEach then we can move them. Does that sound reasonable?

Reply:
I worry a bit that users might not know they have to manually clean up failed workload clusters on a failed test. Would it be better to expose a cli flag or use an env var to avoid deleting the workload cluster (and possibly the kind cluster hosting the management cluster) for debugging purposes?

Reply:
that's a good suggestion, it would be better, let me play around with that today

Reply:
As of now the behavior when a test fails the kind cluster `mgmt` gets cleaned up so it doesn't really help with troubleshooting. I'm ok with passing a flag that keeps the resources around for troubleshooting purposes but I also don't think it's a blocker. It could be a good separate issue to work on maybe.

Reply:
ok. I'm going to add this as follow up work to move things along